sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala

```diff
@@ -442,3 +442,191 @@ case class ArrayAggregate(
+  """,
+  since = "2.4.0")
+case class MapZipWith(left: Expression, right: Expression, function: Expression)
+  extends HigherOrderFunction with CodegenFallback {
+
+  @transient lazy val functionForEval: Expression = functionsForEval.head
+
+  @transient lazy val (keyType, leftValueType, _) =
+    HigherOrderFunction.mapKeyValueArgumentType(left.dataType)
+
+  @transient lazy val (_, rightValueType, _) =
+    HigherOrderFunction.mapKeyValueArgumentType(right.dataType)
+
+  @transient lazy val ordering = TypeUtils.getInterpretedOrdering(keyType)
+
+  override def inputs: Seq[Expression] = left :: right :: Nil
+
+  override def functions: Seq[Expression] = function :: Nil
+
+  override def nullable: Boolean = left.nullable || right.nullable
+
+  override def dataType: DataType = MapType(keyType, function.dataType, function.nullable)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    (left.dataType, right.dataType) match {
+      case (MapType(k1, _, _), MapType(k2, _, _)) if k1.sameType(k2) =>
+        TypeUtils.checkForOrderingExpr(k1, s"function $prettyName")
+      case _ => TypeCheckResult.TypeCheckFailure(s"The input to function $prettyName should have " +
+        s"been two ${MapType.simpleString}s with the same key type, but it's " +
+        s"[${left.dataType.catalogString}, ${right.dataType.catalogString}].")
+    }
+  }
+
+  override def bind(f: (Expression, Seq[(DataType, Boolean)]) => LambdaFunction): MapZipWith = {
+    val arguments = Seq((keyType, false), (leftValueType, true), (rightValueType, true))
+    copy(function = f(function, arguments))
+  }
+
+  override def eval(input: InternalRow): Any = {
+    val value1 = left.eval(input)
+    if (value1 == null) {
+      null
+    } else {
+      val value2 = right.eval(input)
+      if (value2 == null) {
+        null
+      } else {
+        nullSafeEval(input, value1, value2)
+      }
+    }
+  }
+
+  @transient lazy val LambdaFunction(_, Seq(
+    keyVar: NamedLambdaVariable,
+    value1Var: NamedLambdaVariable,
+    value2Var: NamedLambdaVariable),
+    _) = function
+
+  private def keyTypeSupportsEquals = keyType match {
+    case BinaryType => false
+    case _: AtomicType => true
+    case _ => false
+  }
+
+  @transient private lazy val getKeysWithValueIndexes:
+      (ArrayData, ArrayData) => Seq[(Any, Array[Option[Int]])] = {
+    if (keyTypeSupportsEquals) {
+      getKeysWithIndexesFast
+    } else {
+      getKeysWithIndexesBruteForce
+    }
+  }
+
+  private def assertSizeOfArrayBuffer(size: Int): Unit = {
+    if (size > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+      throw new RuntimeException(s"Unsuccessful try to zip maps with $size " +
+        s"unique keys due to exceeding the array size limit " +
+        s"${ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH}.")
+    }
+  }
+
+  private def getKeysWithIndexesFast(keys1: ArrayData, keys2: ArrayData) = {
+    val arrayBuffer = new mutable.ArrayBuffer[(Any, Array[Option[Int]])]
```

Comment:
If we changed it to ```(Option[Int], Option[Int])```, wouldn't we need two similar ```i``` loops instead of one?

My motivation for using also the ```ArrayBuffer``` is preserve the order of keys. A random order would break map comparison in tests. Maybe you will come with idea how to compare maps in tests better :-)

Reply:
> If we changed it to (Option[Int], Option[Int]), wouldn't we need two similar i loops instead of one?
I really don't think so, it would be the same as now I think

well, maybe we can fix map comparison in tests... :)

Reply:
> I really don't think so, it would be the same as now I think

Let's assume that ```indexes``` are tuple for now. ```indexes(z).isEmpty``` could replace with ```indexes.productElement(z).isEmpty```, but how to replace ```indexes(z) = Some(i)```? Since tuple is immutable, I don't see how to replace ith element with ```copy``` function. Maybe we could implement a dedicated class to hold indexes, but is it worth doing that?

Reply:
since the HashMap is mutable, you can just: `hashMap += key -> newTuple`

Reply:
you can change a bit approach from the current one. For the first array of keys, in particular, you don't need to check neither whether the key is there nor the size of the output array, you just need to add them. Then you can add the keys from the other one with the logic here.

Reply:
> you don't need to check neither whether the key is there nor the size of the output array, you just need to add them.

What about duplicated keys? They can be created with other map functions.

Reply:
That is a problem anyway also with the current implementation, as you would pick only one value in that case. The same if you add it without checking.

Reply:
we can the same check, I don't think it is really needed as having duplicated keys is anyway a bad situation itself which shouldn't happen.

Reply:
This is valid argument, it's a rare edge case. The last question before I change it. WDYT about performance a mutable array vs. ```oldTuple.copy(_2 = newValue)```?

Reply:
The array based solution is 20% faster indeed according to my benchmark, but I think it is not critical as I run the benchmark performing the operation 1.000.000 times and the absolute difference was 2 ms. So I prefer the cleaner solution (that is using tuples). @ueshin what do you think?

Reply:
I generally prefer the cleaner solution, but actually I'd prefer the previous approach in this case for 2 reasons:

- We shouldn't ignore 20% of performance difference.
- I'm not sure we can modify the comparison of `MapType` in `ExpressionEvalHelper` here. We might need another pr to make sure the modification is valid.

We still need to add comments what the arrays are for and the reason, though.

Reply:
@mgaido91 Are you comfortable with reverting back to the previous version?

Reply:
Like this idea, thanks!